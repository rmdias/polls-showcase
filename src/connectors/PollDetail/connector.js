import _ from 'lodash'
import { connect } from 'react-redux'
import { poll } from 'actions/poll'

export default connect(
  (state, ownProps) => ({
    poll: _.get(state, 'poll.data', {}),
    loading: _.get(state, 'poll.loading', true)
  }),
  (dispatch) => ({
    onVote: choiceId => dispatch(poll.update.request(choiceId)),
    onFetchPoll: questionId => dispatch(poll.fetch.request(questionId))
  })
)
