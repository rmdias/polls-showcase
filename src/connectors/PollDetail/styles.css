.poll {
  width: 100%;
  margin: 0 auto;
  max-width: 700px;
  padding-top: 40px;
  padding-bottom: 70px;
}

.back_to_polls {
  height: 40px;
  color: #333;
  right: 0;
  opacity: 0;
  position: fixed;
  border-radius: 2px;
  display: inline-block;
  text-decoration: none;
  background: #fff;
  padding: 12px 30px 11px 15px;
  box-shadow: 0 0px 6px 0 rgba(0, 0, 0, 0.1);
  transform: translateX(100%) translateY(-90%) translateZ(0);
  animation: slidein 750ms cubic-bezier(0.1, 0.34, 0.53, 1.43) forwards;
}

.poll__choices {
  padding-bottom: 30px;
}

.poll__voting_button {
  float: right;
}

@keyframes slidein {
  from {
    opacity: 0;
    transform: translateX(100%) translateY(-90%);
  }

  to {
    opacity: 1;
    transform: translateX(0) translateY(-90%);
  }
}
