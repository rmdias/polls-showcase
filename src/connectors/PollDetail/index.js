import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import { withRouter } from 'react-router'

import { CircularLoader, Button } from 'luna-ui-lib'

// Components
import ChoiceItem from 'components/ChoiceItem'

import './styles.css'
import connector from './connector'

class PollDetail extends PureComponent {
  static propTypes = {
    loading: PropTypes.bool,
    poll: PropTypes.object,
    onVote: PropTypes.func,
    onFetchPoll: PropTypes.func
  }

  constructor(props) {
    super(props)

    window.scrollTo(0, 0)
    this.state = { selectedChoice: '' }
  }

  handleVotingSelection(choice) {
    if (choice === this.state.selectedChoice) {
      return this.setState((state, props) => ({ selectedChoice: '' }))
    }

    this.setState((state, props) => ({ selectedChoice: choice }))
  }

  vote() {
    this.props.onVote(this.state.selectedChoice)

    setTimeout(() => this.setState((state, props) => ({ selectedChoice: '' })), 500)
  }

  isSelectedChoice(choice) {
    return this.state.selectedChoice === choice.url
  }

  componentDidMount() {
    this.props.onFetchPoll(this.props.match.params.questionId)
  }

  render() {
    const showLoader = this.props.loading && !this.props.poll.question
    const loader = <CircularLoader className="polls__loader" key={0} kind="primary" size={50} border={1} />
    const totalVotes = this.props.poll.choices && this.props.poll.choices.reduce((acc, value) => acc += value.votes, 0)
    const biggerNumber = this.props.poll.choices && this.props.poll.choices.sort((a, b) => a.votes + b.votes)[0].votes

    const Items = this.props.poll.choices && this.props.poll.choices.map(choice => {
      return <ChoiceItem
        key={choice.url}
        choice={choice}
        active={this.isSelectedChoice(choice)}
        totalvotes={totalVotes}
        biggernumber={biggerNumber}
        onClick={() => this.handleVotingSelection(choice.url)}
      />
    })

    return (
      <div className="poll">
        <Link className="back_to_polls" to="/">â—„ Back</Link>
        <h1>Question: {this.props.poll.question}</h1>
        { showLoader && loader }

        <div className="poll__choices">
          { Items }
        </div>

        { this.state.selectedChoice && <Button
          kind="primary"
          className="poll__voting_button"
          promise={this.props.loading}
          config={{
            onClick: () => this.vote()
          }}
          >
            save vote
          </Button>
        }
      </div>
    )
  }
}

export default connector(withRouter(PollDetail))
