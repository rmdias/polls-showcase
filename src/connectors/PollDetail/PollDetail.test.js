import React from 'react'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { BrowserRouter as Router } from 'react-router-dom'

const middlewares = []
const mockStore = configureStore(middlewares)

import PollDetail from './index'
import PollMock from './__mocks__/__mock__poll'

import { Theme } from 'luna-ui-lib'

const poll = {
  data: PollMock,
  loading: false
}
const initialState = { poll }
const store = mockStore(initialState)

beforeEach(() => {
  global.scrollTo = jest.fn()
})

describe('<PollDetail />', () => {
  it('toMatchSnapshot: Loading', () => {
    const component = (
      <Theme>
        <Router>
          <PollDetail store={store}/>
        </Router>
      </Theme>
    )

    const tree = renderer
      .create(component)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

})
