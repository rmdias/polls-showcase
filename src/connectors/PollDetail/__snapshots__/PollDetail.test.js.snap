// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PollDetail /> toMatchSnapshot: Loading 1`] = `
<div>
  <div
    className="poll"
  >
    <a
      className="back_to_polls"
      href="/"
      onClick={[Function]}
    >
      â—„ Back
    </a>
    <h1>
      Question: 
      Game Genre
    </h1>
    <div
      className="poll__choices"
    >
      <div
        onClick={[Function]}
      >
        <div
          className="poll__choice_item false sc-kpOJdX ddRiWz"
        >
          <div>
            <strong>
              Strategy
            </strong>
          </div>
          <div>
            270
             votes
          </div>
          <div>
            14
            %
          </div>
          <div
            className="poll__choice_percentage"
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "width": 200,
                }
              }
            >
              <div
                style={
                  Object {
                    "backgroundColor": "#b2e2bf",
                    "fontSize": undefined,
                    "height": "16px",
                    "padding": "1px 2px",
                    "width": 100,
                  }
                }
              />
              <div
                style={
                  Object {
                    "backgroundColor": "#c7c7c7",
                    "fontSize": undefined,
                    "height": "16px",
                    "padding": "1px 2px",
                    "width": 100,
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
      <div
        onClick={[Function]}
      >
        <div
          className="poll__choice_item false sc-dxgOiQ kRJcgL"
        >
          <div>
            <strong>
              Action
            </strong>
          </div>
          <div>
            409
             votes
          </div>
          <div>
            21
            %
          </div>
          <div
            className="poll__choice_percentage"
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "width": 200,
                }
              }
            >
              <div
                style={
                  Object {
                    "backgroundColor": "#b2e2bf",
                    "fontSize": undefined,
                    "height": "16px",
                    "padding": "1px 2px",
                    "width": 120.47128129602356,
                  }
                }
              />
              <div
                style={
                  Object {
                    "backgroundColor": "#c7c7c7",
                    "fontSize": undefined,
                    "height": "16px",
                    "padding": "1px 2px",
                    "width": 79.52871870397644,
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
      <div
        onClick={[Function]}
      >
        <div
          className="poll__choice_item false sc-ckVGcZ EUALb"
        >
          <div>
            <strong>
              Shooter
            </strong>
          </div>
          <div>
            1195
             votes
          </div>
          <div>
            63
            %
          </div>
          <div
            className="poll__choice_percentage"
          >
            <div
              style={
                Object {
                  "display": "flex",
                  "width": 200,
                }
              }
            >
              <div
                style={
                  Object {
                    "backgroundColor": "#b2e2bf",
                    "fontSize": undefined,
                    "height": "16px",
                    "padding": "1px 2px",
                    "width": 163.13993174061434,
                  }
                }
              />
              <div
                style={
                  Object {
                    "backgroundColor": "#c7c7c7",
                    "fontSize": undefined,
                    "height": "16px",
                    "padding": "1px 2px",
                    "width": 36.86006825938566,
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    
  </div>
</div>
`;
